wfe/wfe.go

```diff
@@ -2565,3 +2584,125 @@ func (wfe *WebFrontEndImpl) processRevocation(
+// then error.
+func (wfe *WebFrontEndImpl) verifyEAB(newAcctReq newAccountRequest,
+	outerPostData *authenticatedPOST) (*acme.JSONSigned, *acme.ProblemDetails) {
+	if newAcctReq.ExternalAccountBinding == nil {
+		if wfe.requireEAB {
+			return nil, acme.ExternalAccountRequiredProblem(
+				"ACME server policy requires newAccount requests must include a value for the 'externalAccountBinding' field")
+		}
+
+		return nil, nil
+	}
+
+	//1.  Verify that the value of the field is a well-formed JWS
+	eabBytes, err := json.Marshal(newAcctReq.ExternalAccountBinding)
+	if err != nil {
+		return nil, acme.MalformedProblem(
+			fmt.Sprintf("failed to decode external account binding: %s", err))
+	}
+
+	eab, err := jose.ParseSigned(string(eabBytes))
+	if err != nil {
+		return nil, acme.MalformedProblem(
+			fmt.Sprintf("failed to decode external account binding: %s", err))
+	}
+
+	//2.  Verify that the JWS protected field meets the following criteria
+	//-  The "alg" field MUST indicate a MAC-based algorithm
+	//-  The "kid" field MUST contain the key identifier provided by the CA
+	//-  The "nonce" field MUST NOT be present
+	//-  The "url" field MUST be set to the same value as the outer JWS
+	keyID, prob := wfe.verifyEABPayloadHeader(eab, outerPostData)
+	if prob != nil {
+		return nil, prob
+	}
+
+	//3.  Retrieve the MAC key corresponding to the key identifier in the
+	//    "kid" field
+	key, ok := wfe.db.GetExtenalAccountKeyByID(keyID)
+	if !ok {
+		return nil, acme.UnauthorizedProblem(
+			"the field 'kid' references a key that is not known to the ACME server")
+	}
+
+	//4.  Verify that the MAC on the JWS verifies using that MAC key
+	payload, err := eab.Verify(key)
+	if err != nil {
+		return nil, acme.UnauthorizedProblem(
+			fmt.Sprintf("external account binding JWS verification error: %s", err))
+	}
+
+	//5.  Verify that the payload of the JWS represents the same key as was
+	//    used to verify the outer JWS (i.e., the "jwk" field of the outer
+	//    JWS)
+	if prob := wfe.verifyEABMatchesKey(payload, outerPostData.jwk); prob != nil {
+		return nil, prob
+	}
+
+	wfe.log.Printf("Successful newAccount Binding with CA using kid %q", keyID)
+
+	return newAcctReq.ExternalAccountBinding, nil
+}
+
+// verifyEABPayloadHeader will verify the protected header object of the
+// External Account Binding object in a newAccount request. If successful, will
+// return the KID specified in the header.
+func (wfe *WebFrontEndImpl) verifyEABPayloadHeader(innerJWS *jose.JSONWebSignature, outerPostData *authenticatedPOST) (string, *acme.ProblemDetails) {
+	if len(innerJWS.Signatures) != 1 {
+		return "", acme.MalformedProblem(
+			"expecting single signature associated with the external account binding")
+	}
+
+	header := innerJWS.Signatures[0].Protected
+	switch header.Algorithm {
+	case "HS256", "HS384", "HS512":
+		break
+	default:
+		return "", acme.BadPublicKeyProblem(
+			"the 'alg' field is not valid for external account binding")
```

Comment:
Can this error message be updated to include the value of the alg field that was rejected and the accepted values?